File: backend/src/main/java/org/luxons/sevenwonders/game/data/serializers/ProductionIncreaseSerializer.java
package org.luxons.sevenwonders.game.data.serializers;

import java.lang.reflect.Type;

import org.luxons.sevenwonders.game.effects.ProductionIncrease;
import org.luxons.sevenwonders.game.resources.Production;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

public class ProductionIncreaseSerializer
        implements JsonSerializer<ProductionIncrease>, JsonDeserializer<ProductionIncrease> {

    @Override
    public JsonElement serialize(ProductionIncrease productionIncrease, Type typeOfSrc,
            JsonSerializationContext context) {
        Production production = productionIncrease.getProduction();
        JsonElement json = context.serialize(production);
        if (!json.isJsonNull() && !productionIncrease.isSellable()) {
            return new JsonPrimitive(wrapInBrackets(json.getAsString()));
        }
        return json;
    }

    private String wrapInBrackets(String str) {
        return '(' + str + ')';
    }

    @Override
    public ProductionIncrease deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
            throws JsonParseException {
        ProductionIncrease productionIncrease = new ProductionIncrease();

        String resourcesStr = json.getAsString();
        boolean isSellable = !resourcesStr.startsWith("(");
        if (!isSellable) {
            resourcesStr = unwrapBrackets(resourcesStr);
            json = new JsonPrimitive(resourcesStr);
        }
        productionIncrease.setSellable(isSellable);

        Production production = context.deserialize(json, Production.class);
        productionIncrease.setProduction(production);
        return productionIncrease;
    }

    private static String unwrapBrackets(String str) {
        return str.substring(1, str.length() - 1);
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/data/serializers/ProductionIncreaseSerializer.kt
package org.luxons.sevenwonders.game.data.serializers

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import org.luxons.sevenwonders.game.effects.ProductionIncrease
import org.luxons.sevenwonders.game.resources.Production
import java.lang.reflect.Type

class ProductionIncreaseSerializer : JsonSerializer<ProductionIncrease>, JsonDeserializer<ProductionIncrease> {

    override fun serialize(
        productionIncrease: ProductionIncrease,
        typeOfSrc: Type,
        context: JsonSerializationContext
    ): JsonElement {
        val production = productionIncrease.production
        val json = context.serialize(production)
        return if (!json.isJsonNull && !productionIncrease.isSellable) { JsonPrimitive("(${json.asString})")} else json
    }

    @Throws(JsonParseException::class)
    override fun deserialize(
        json: JsonElement,
        typeOfT: Type,
        context: JsonDeserializationContext
    ): ProductionIncrease {
        var json = json
        val productionIncrease = ProductionIncrease()

        var resourcesStr = json.asString
        val isSellable = !resourcesStr.startsWith("(")
        if (!isSellable) {
            resourcesStr = unwrapBrackets(resourcesStr)
            json = JsonPrimitive(resourcesStr)
        }
        productionIncrease.isSellable = isSellable

        val production = context.deserialize<Production>(json, Production::class.java)
        productionIncrease.production = production
        return productionIncrease
    }

    private fun unwrapBrackets(str: String): String {
        return str.substring(1, str.length - 1)
    }
}
